### PR TITLE
Support routing by encoding

### PR DESCRIPTION
**Pull request by kriskowal**
This change makes it possible to route to alternate procedures with the same
service name and procedure name, but alternate encodings.
For example, this would enable a service to implement "echo" with "raw" and
"json" alternatives.
This would also allow the thrift or protobuf encodings to register a "json"
variant of every handler using the same models.

I've provided this change to give @peter-edge some direction for gRPC JSON
gateway support.

### CODE REVIEW

**Comment by kriskowal** [comment]
Taking a note from offline: this also makes it possible for the Thrift "encoding" to register "thrift" and "thrifte" (with envelope) alternatives, or "tbinary" and "tcompact" alternatives. Effectively what we call "encoding" is a type model on one hand and a collection of supported wire encodings on the other.

File: router.go [hunk 1]
```diff
@@ -63,39 +72,76 @@ func (m MapRouter) Register(rs []transport.Procedure) {
...
+		// This supports wild card encodings (for backward compatibility,
+		// since type models like Thrift were not previously required to
+		// specify the encoding of every procedure)
 		sp := serviceProcedure{
 			service:   r.Service,
 			procedure: r.Name,
 		}
-		m.entries[sp] = r
+		// Protect against registering ambiguous wildcards
+		if _, ok := m.serviceProcedures[sp]; r.Encoding == "" && ok {
```
Comment (peter-edge) [inline router.go]: Do you also want to check if this is in `m.serviceProcedureEncodings`? I think we should outlaw registering a `transport.Procedure` with an encoding, and without an encoding - there should either be one `transport.Procedure` without an encoding, or >=1 `transport.Procedure`s with an encoding.

**Comment by peter-edge** [comment]
I merged dev onto this so I can try developing proto changes on top of it

File: api/transport/router.go [hunk 2]
```diff
@@ -59,8 +60,11 @@ func (p Procedure) MarshalLogObject(enc zapcore.ObjectEncoder) error {
 	return enc.AddObject("handler", p.HandlerSpec)
 }
 
-// Less orders procedures lexicographically on (Service, Name).
+// Less orders procedures lexicographically on (Service, Name, Encoding).
 func (p Procedure) Less(other Procedure) bool {
+	if p.Name == other.Name {
```
Comment (willhug) [inline api/transport/router.go]: this condition ignores the service... so if I have two services with the same schema it will fail to sort properly.

File: router.go [hunk 3]
```diff
@@ -63,39 +72,76 @@ func (m MapRouter) Register(rs []transport.Procedure) {
 			panic("Expected procedure name not to be empty string in registration")
 		}
 
+		if r.Encoding != "" {
```
Comment (willhug) [inline router.go]: why don't we just register everything in the encoding list?  That way we have a single source of truth for *all* procedures.

File: router.go [hunk 4]
```diff
@@ -63,39 +72,76 @@ func (m MapRouter) Register(rs []transport.Procedure) {
...
 // Procedures returns a list procedures that
 // have been registered so far.
 func (m MapRouter) Procedures() []transport.Procedure {
-	procs := make([]transport.Procedure, 0, len(m.entries))
-	for _, v := range m.entries {
+	procs := make([]transport.Procedure, 0, len(m.serviceProcedures)+len(m.serviceProcedureEncodings))
+	for _, v := range m.serviceProcedureEncodings {
+		procs = append(procs, v)
+	}
+	for _, v := range m.serviceProcedures {
```
Comment (willhug) [inline router.go]: this will return duplicate procedures now...

Comment (kriskowal) [inline router.go]: Addressed in #1053 which makes (service, procedure) and (service, procedure, encoding) routing tables mutually exclusive.

File: router.go (hunk 3 above)
Comment (kriskowal) [inline router.go]: Supported encodings is open-ended, so we have to treat a blank encoding as a wild card. We do not have a list of supported encodings to enumerate to fill the holes in the (service, procedure, encoding) table if an encoding is not specified.

Comment (willhug) [inline router.go]: oh, i was thinking of having this function in tandem with the "Procedures" function below, I see how it got misconstrued with actually routing to procedures (which wouldn't work for the reasons you specified).

File: api/transport/router.go (hunk 2 above)
Comment (kriskowal) [inline api/transport/router.go]: Addressed in #1054

File: router.go (hunk 1 above)
Comment (kriskowal) [inline router.go]: Addressing in #1053

File: api/transport/router.go [hunk 5]
```diff
@@ -40,7 +40,8 @@ type Procedure struct {
 	// HandlerSpec specifiying which handler and rpc type.
 	HandlerSpec HandlerSpec
 
-	// Encoding of the handler, for introspection.
+	// Encoding of the handler, optional, used for introspection, and used for
+	// routing if present.
```
Comment (abhinav) [inline api/transport/router.go]: More sentence, less list please.

> Encoding of the handler. This field is optional.

We don't need to mention exactly what it's used for because that list can expand (as it has already in this change).

File: api/transport/router.go [hunk 6]
```diff
@@ -59,12 +60,30 @@ func (p Procedure) MarshalLogObject(enc zapcore.ObjectEncoder) error {
...
 	}
-	return p.Service < other.Service
+	if p.Name != o.Name {
+		return p.Name < o.Name
+	}
+	return p.Encoding < o.Encoding
+}
+
+// Procedures is a sortable slice of procedures.
+type Procedures []Procedure
```
Comment (abhinav) [inline api/transport/router.go]: Adding this to the public API is not needed. This is used in only one place. Unnecessary added API surface.

File: router.go [hunk 7]
```diff
@@ -36,19 +37,27 @@ type serviceProcedure struct {
...
+}
+
 // MapRouter is a Router that maintains a map of the registered
 // procedures.
 type MapRouter struct {
-	defaultService string
-	entries        map[serviceProcedure]transport.Procedure
+	defaultService            string
+	serviceProcedures         map[serviceProcedure]transport.Procedure
+	serviceProcedureEncodings map[serviceProcedureEncoding]transport.Procedure
```
Comment (abhinav) [inline router.go]: This seems a bit convoluted. Perhaps it will be easier to do,

    map[serviceProcedure]encodingProcedure

Where

    type encodingProcedure interface {
        GetProcedure(transport.Encoding) transport.Procedure
    }

With two implementations,

    type wildcardEncodingProcedure struct{ transport.Procedure }
    type perEncodingProcedure map[transport.Encoding]transport.Procedure

This also makes supportedEncodings unnecessary.

You can add other helper functions like `ListProcedures() []transport.Procedure` (for `Procedures()`) as needed.

File: api/transport/errors.go [hunk 8]
```diff
@@ -20,7 +20,9 @@
 
 package transport
 
-import "go.uber.org/yarpc/internal/errors"
+import (
+	"go.uber.org/yarpc/internal/errors"
+)
```
Comment (abhinav) [inline api/transport/errors.go]: nit: no groups for single imports

File: internal/errors/brands_test.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+import "testing"
+
+func TestCoverBrands(t *testing.T) {
+	// sorted
+	RemoteTimeoutError("").timeoutError()
+	clientTimeoutError{}.clientError()
+	clientTimeoutError{}.timeoutError()
+	handlerBadRequestError{}.badRequestError()
+	handlerBadRequestError{}.handlerError()
+	handlerTimeoutError{}.handlerError()
```
Comment (abhinav) [inline internal/errors/brands_test.go]: should error brands be their own PR?

File: internal/joinencodings/joinencodings.go [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package joinencodings
+
+import "fmt"
+
+// Join joins a list of encodings as a English (Chicago Style Manual)
+// string for presentation in diagnostic messages.
+func Join(encs []string) string {
```
Comment (abhinav) [inline internal/joinencodings/joinencodings.go]: Maybe rename package to "oxfordcomma" and the function to `Or`. The "no encodings" would have to be an argument.

    package oxfordcomma

    func Or(pluralNoun string, items []string) string { ... }

I'm certain we do the "and" form manually elsewhere so that can be added for a `func Add`.

Otherwise this package seems wholly unnecessary and too specific to this case, and it can just be a private function where it's used.

File: router.go [hunk 11]
```diff
@@ -36,23 +38,41 @@ type serviceProcedure struct {
...
+		serviceProcedures:         make(map[serviceProcedure]transport.Procedure),
+		serviceProcedureEncodings: make(map[serviceProcedureEncoding]transport.Procedure),
+		supportedEncodings:        make(map[serviceProcedure][]string),
 	}
 }
 
 // Register registers the procedure with the MapRouter.
+// If the procedure does not specify its service name, the procedure will
+// inherit the default service name of the router.
+// Procedures should specify their encoding, and multiple procedures with the
```
Comment (abhinav) [inline router.go]: s/should/may? (RFC2119)

File: router.go [hunk 12]
```diff
@@ -36,23 +38,41 @@ type serviceProcedure struct {
...
+// same name and service name can exist if they handle different encodings.
+// However, specifying the encoding is optional since it was not required
+// in version 1.
```
Comment (abhinav) [inline router.go]: This is still version 1. I would drop this comment.

File: internal/joinencodings/joinencodings.go (hunk 10 above)
Comment (kriskowal) [inline internal/joinencodings/joinencodings.go]: This function is used in two packages and trying to implement it in one or the other caused cyclic imports, since the yarpc root package is both API and implementation. yarpc.CallOption is API for the internal/encodings package, and I need to use this from the yarpc package.

File: router.go (hunk 12 above)
Comment (kriskowal) [inline router.go]: 1.0.0?

File: api/transport/router.go (hunk 6 above)
Comment (kriskowal) [inline api/transport/router.go]: Making this public facilitated coverage over the Less/Swap&c. Do we still like coverage?

File: internal/errors/brands_test.go (hunk 9 above)
Comment (kriskowal) [inline internal/errors/brands_test.go]: I needed one of these to get coverage on the brand I added. Seems like an inefficient use of time to make a separate PR.

File: router.go (hunk 7 above)
Comment (kriskowal) [inline router.go]: For the sake of consideration, we could put Register on the encodingProcedure implementation to handle the conflicting cases (where encoding is specified and a wildcard encoding is already in place, where an encoding is omitted where a specific encoding is already in place).

The downside is that it adds an interface indirection where it’s not necessary to facilitate polymorphism for the user. Ideally we can shape this code so removing support for `serviceProcedures` is strictly deleting some lines, not refactoring.

File: api/transport/router.go (hunk 6 above)
Comment (abhinav) [inline api/transport/router.go]: It doesn't need to be public for coverage. A test in the same package as where this is used will have access to the private `byProcedures` type. We shouldn't expose new APIs for the sake of coverage.

File: router.go (hunk 12 above)
Comment (abhinav) [inline router.go]: I mean, the contract at this point is that encoding is not required. This is essentially saying that "encoding isn't required because it wasn't required". We can simply say that specifying the encoding is recommended but optional (and so and so happens when you specify it and when you don't).

File: router.go (hunk 11 above)
Comment (kriskowal) [inline router.go]: def SHOULD

File: router.go (hunk 7 above)
Comment (abhinav) [inline router.go]: A single `map[serviceProcedure]transport.Procedure` is just analog for `map[service]map[procedure]transport.Procedure`. Two maps make this a lot more complicated than it needs to be. It's cleaner/simpler/more straightforward to have a single map than have two maps and somehow enforce the invariant that a key exists in only one of the two maps. Plus that's two map lookups (in the worst case) at dispatch rather than one.

RE: Interface Indirection: Just an implementation detail. Interface is just one way to do it. Another is a union:

    // With map[serviceProcedure]encodingProcedure,
    type encodingProcedure struct {
        // Only one of the following is set.
        Wildcard transport.Procedure
        PerEncoding map[transport.Encoding]transport.Procedure
    }

File: internal/joinencodings/joinencodings.go (hunk 10 above)
Comment (abhinav) [inline internal/joinencodings/joinencodings.go]: I mean I noticed that it's used in two places. That's why I suggested generalizing this package rather than keeping it this specific. I guess I'm just trying to avoid the extremely specific package with a single function. This package's name suggests it can only ever have one function: "Join(encodings)".

File: router.go (hunk 7 above)
Comment (kriskowal) [inline router.go]: Agree that it’s two look ups in the legacy case, but it’s only one in the common case, and the common case is the only case that matters. It is the only case that will remain when we remove the legacy case and in practice is the only case that will be exercised.

File: internal/joinencodings/joinencodings.go (hunk 10 above)
Comment (kriskowal) [inline internal/joinencodings/joinencodings.go]: Agreed. I think I will call it `humanize` in the grand tradition and generalize the method to`Join(list, conjunction, none)`.

**Comment by kriskowal** [comment]
@willhug I’ll chase you for a follow-up review on Monday :)

File: internal/errors/router.go [hunk 13]
```diff
@@ -56,3 +60,37 @@ func (e unrecognizedProcedureError) Error() string {
...
+	}
+}
+
+type unrecognizedEncodingError struct {
+	Want []string
+	Got  string
+}
+
+// brands this uniquely as an unrecognized encoding error.
+func (unrecognizedEncodingError) unrecognizedEncodingError() {}
```
Comment (willhug) [inline internal/errors/router.go]: should an unrecognizedEncoding error be a subset of the UnrecognizedProcedureError?

File: router.go [hunk 14]
```diff
@@ -67,51 +85,123 @@ func (m MapRouter) Register(rs []transport.Procedure) {
...
+		service:   service,
+		procedure: procedure,
+	}
+	if procedure, ok := m.serviceProcedures[sp]; ok {
+		return procedure.HandlerSpec, nil
+	}
 
-type proceduresByServiceProcedure []transport.Procedure
+	// Supported procedure, unrecognized encoding.
+	if want, ok := m.supportedEncodings[sp]; ok {
```
Comment (willhug) [inline router.go]: `s/want/supportedEncodings`?

File: router.go [hunk 15]
```diff
@@ -67,51 +85,123 @@ func (m MapRouter) Register(rs []transport.Procedure) {
 			service:   r.Service,
 			procedure: r.Name,
 		}
-		m.entries[sp] = r
+
+		if r.Encoding == "" {
+			// Protect against masking encoding-specific routes.
+			if _, ok := m.serviceProcedures[sp]; ok {
+				panic(fmt.Sprintf("Cannot register multiple handlers for every encoding for service %q and procedure  %q", sp.service, sp.procedure))
```
Comment (peter-edge) [inline router.go]: I strongly disagree - why would someone want to override a procedure? This will mess with a lot of logic and is effectively a bug

File: router.go (hunk 14 above)
Comment (peter-edge) [inline router.go]: `want` is used in a lot of places in the codebase, why are we reviewing this?

File: router.go (hunk 11 above)
Comment (peter-edge) [inline router.go]: I don't think we should be reviewing comment wording

File: internal/humanize/join_test.go [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+package humanize
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestQuotedJoin(t *testing.T) {
+	tests := []struct {
```
Comment (peter-edge) [inline internal/humanize/join_test.go]: nit: just putting this out there, this is an insanely complicated way to do:

```go
assert.Equal(t, `"none"`, QuotedJoin(nil, "or", "none"))
assert.Equal(t, `"json"`, QuotedJoin([]string{"json"}, "or", "none"))
assert.Equal(t, `"json" or "proto"`, QuotedJoin([]string{"json", "proto"}, "or", "none"))
assert.Equal(t, `"json", "thrift", or "proto"`, QuotedJoin([]string{"json", "thrift", "proto"}, "or", "none"))
```

We don't have to do this style for every test

File: internal/errors/router.go [hunk 17]
```diff
@@ -56,3 +60,37 @@ func (e unrecognizedProcedureError) Error() string {
...
+
+func (e unrecognizedEncodingError) Error() string {
```
Comment (peter-edge) [inline internal/errors/router.go]: @kriskowal @willhug why do we do these errors as structs, not struct pointers? it means we are copying a pointer and a string ref

File: router.go (hunk 14 above)
Comment (willhug) [inline router.go]: Because code is read more than it is written.  My philosophy is to bring up anything that slowed down my understanding of the code if I can offer an alternative.

In this particular case, I jumped to the `len(want) == 1` and scratched my head a bit because we start treating this want as a list, which is not insinuated at all by the variables naming, which is confusing.

Comment (peter-edge) [inline router.go]: Then we should file an issue and fix this kind of thing across the whole codebase, we need to be consistent. I'll fix it this time but this is the kind of stuff we shouldn't be worrying about IMO.

File: api/transport/errors.go [hunk 18]
```diff
@@ -63,3 +63,18 @@ func IsUnrecognizedProcedureError(err error) bool {
...
+// IsUnrecognizedEncodingError can distinguish it from other errors coming out
+// of router.Choose.
+func UnrecognizedEncodingError(req *Request, want []string) error {
+	return errors.RouterUnrecognizedEncodingError(want, string(req.Encoding))
+}
+
+// IsUnrecognizedEncodingError returns true for errors returned by
+// Router.Choose if the router cannot find a handler for the request's
+// encoding.
+func IsUnrecognizedEncodingError(err error) bool {
```
Comment (peter-edge) [inline api/transport/errors.go]: Do we really need these? This is expanding on an API that we're about to do work on with the error stuff @kriskowal

File: internal/humanize/join.go [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+package humanize
+
+import "fmt"
+
+// QuotedJoin transforms a list of terms into an English (Chicago Manual Style)
+// serial comma delimited list with the given conjunction ("and", "or") and
+// each term enquoted.
+// It displays the "none" case if the list is empty, like "no soup".
+func QuotedJoin(terms []string, conjunction, none string) string {
```
Comment (peter-edge) [inline internal/humanize/join.go]: This seems like overkill @kriskowal

File: router.go (hunk 15 above)
Comment (willhug) [inline router.go]: offline talk, technically a breaking change, we'll go forward with the panic case here, if anyone complains peter will give me 10$.

File: internal/errors/router.go (hunk 17 above)
Comment (willhug) [inline internal/errors/router.go]: (guess) I would assume it's so we avoid heap allocation.  I don't know enough about the tradeoffs there though, one heap allocation vs a bunch of larger stack allocations.

**Comment by peter-edge** [comment]
When you pass structs as parameters to functions it allocates a copy on the
heap anyways

**Comment by willhug** [comment]
Turns out UnrecognizedProcedureError messages are enforced as a part of the wire protocol, which is slightly insane.  So I'm removing all encoding information from the error.  I want to add the full errors in the future, but there's no way to do that without increasing API exposure or using internal errors.
